src/snowflake/snowpark/mock/_plan.py

```diff
@@ -536,24 +540,36 @@ def cleanup_imports():
         else:
             udf_handler = udf.func
 
-        # Compute
+        # Compute input data and validate typing
+        if len(exp.children) != len(udf._input_types):
+            raise SnowparkLocalTestingException(
+                f"Expected {len(udf._input_types)} arguments, but received {len(exp.children)}"
+            )
+
         function_input = TableEmulator(index=input_data.index)
-        for child in exp.children:
+        for child, expected_type in zip(exp.children, udf._input_types):
             col_name = analyzer.analyze(child, expr_to_alias)
-            function_input[col_name] = calculate_expression(
+            column_data = calculate_expression(
                 child, input_data, analyzer, expr_to_alias
             )
-        # Validate function inputs
-        actual_types = [col.datatype for col in function_input.sf_types.values()]
-        if len(actual_types) != len(udf._input_types):
-            raise SnowparkLocalTestingException(
-                f"Expected {len(udf._input_types)} arguments, but received {len(actual_types)}"
-            )
 
-        for expected, actual in zip(udf._input_types, actual_types):
-            if not types_are_compatible(expected, actual):
+            # SNOW-929218: Once proper type coercion is supported use that instead.
+            if isinstance(expected_type, VariantType) and not isinstance(
```

Comment:
this is beyond the scope of this PR, does we support MapType/ArrayType in UDF and if so do they need to coerce like VariantType?

Reply:
We do support them, but I don't think they need coercion in a similar way to Variant type. I don't think server side would try to implicitly cast an integer to a map in the same way it would an integer into a variant for example.